crates/executor/state_handler/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+pub struct StateHandler<S: Store> {
+    snapshot_state: SnapshotState<S>,
+    /// Checkpoints are used to revert state changes.
+    /// checkpoints are used to follow the call stack of a transaction
+    /// and revert the state changes when the transaction is rolled back.
+    checkpoints: Vec<Snapshot>,
+}
+
+impl<S: Store> StateHandler<S> {
+    /// Create a new state handler with the given store.
+    pub fn new(store: S) -> Self {
+        Self {
+            snapshot_state: SnapshotState::new(store),
+            checkpoints: Vec::with_capacity_in(200, Global),
+        }
+    }
+
+    pub(crate) fn construct_key(
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        accumulator: bool,
+    ) -> Vec<u8> {
+        const KV_SEPARATOR: u8 = 0;
+        const KV_SCOPED_SEPARATOR: u8 = 1;
+        const ACC_SEPARATOR: u8 = 2;
+        const SCOPED_ACC_SEPARATOR: u8 = 3;
+        match scope {
+            // account / 0 / key
+            // account / 1 / scope / key
+            // account / 2 (accumulator) / key
+            // account / 3 (scoped accumulator) / scope / key
+            Some(scope) => {
+                let ac = account_id.to_le_bytes();
+                let sc = scope.to_le_bytes();
+                let mut new_key = Vec::new_in(Global);
+                new_key.extend_from_slice(&ac);
+                if accumulator {
+                    new_key.push(SCOPED_ACC_SEPARATOR);
+                } else {
+                    new_key.push(KV_SCOPED_SEPARATOR);
+                }
+                new_key.extend_from_slice(&sc);
+                new_key.extend_from_slice(key);
+                new_key
+            }
+            None => {
+                let mut new_key = Vec::new_in(Global);
+                new_key.extend_from_slice(&account_id.to_le_bytes());
+                if accumulator {
+                    new_key.push(ACC_SEPARATOR);
+                } else {
+                    new_key.push(KV_SEPARATOR);
+                }
+                new_key.extend_from_slice(key);
+                new_key
+            }
+        }
+    }
+}
+
+impl<S: Store> StdStateManager for StateHandler<S> {
+    fn kv_get<A: Allocator>(
+        &self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        allocator: A,
+    ) -> Result<Option<Vec<u8, A>>, StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, false);
+
+        match self.snapshot_state.get(&constructed_key, allocator) {
+            Some(value) => Ok(Some(value)),
+            None => Ok(None),
+        }
+    }
+
+    fn kv_set(
+        &mut self,
+        account_id: AccountID,
+        scope: Option<AccountID>,
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), StdStateError> {
+        let constructed_key = Self::construct_key(account_id, scope, key, false);
+        let mut vec = Vec::new(); //TODO allocations occur here
```

Comment:
I think in my latest branch I have the Allocator getting passed down into kv_set for this very reason. So we can refactor it after that. We should make a note in linear